sw/otbn/crypto/field25519.s

```diff
@@ -0,0 +1,350 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * This library contains arithmetic for the finite field modulo the prime
+ * p = 2^255-19, which is used for both the X25519 key exchange and the Ed25519
+ * signature scheme.
+ */
+
+
+/**
+ * Fully reduce a 510-bit number modulo p.
+ *
+ * Returns c = a mod p.
+ *
+ * Uses Solinas/pseudo-Mersenne reduction, which is based on the observation
+ * that if a large number x is split into two so that the lower 255 bits are x0
+ * and all bits above 255 are x1, then:
+ *
+ *   x = x0 + (2^255 * x1) \equiv (x0 + (19 * x1)) (mod p)
+ *
+ * If x is large, then x0 + 19 * x1 will be much smaller than x, because 19 <<
+ * 2^255. This step can be repeated as necessary until a conditional
+ * subtraction is enough to fully reduce.
+ *
+ * This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w19: constant, w19 = 19
+ * @param[in]  [w21:w20]: a, number to be reduced (a < 2^510)
+ * @param[in]  MOD: p, modulus = 2^255 - 19
+ * @param[in]  w31: all-zero
+ * @param[out] w22: c, result = a mod p
+ *
+ * clobbered registers: w18, w20 to w22
+ * clobbered flag groups: FG0
+ */
+fe_reduce:
+  /* First Solinas step (reducing modulo 2p = 2^256-38). */
```

Comment:
I don't think we do this anywhere, but do you think it would make sense to post links to either Wikipedia or a relevant cryptographer's blog on the details of these algorithms? Like, I could certainly look this up myself, but I think future readers might appreciate a handy link.

(Maybe it's a thing to do in a single future pass...)

Reply:
Honestly, I _wish_ I knew a reference for a good explanation. Wikipedia has a page [here](https://en.wikipedia.org/wiki/Solinas_prime), although the explanation there seems very overcomplicated to me. I actually just found the Wikipedia page now -- I learned the algorithm by reading assembly code and after some debate called it "Solinas reduction" for the fiat-crypto paper since that's the author of the earliest paper I could find using this method.

Option 1: we rewrite the Wikipedia page
Option 2: Jade starts a blog
Option 3: all of the above

Reply:
I gave a brief explanation in another PR [here](https://github.com/lowRISC/opentitan/pull/11163#issuecomment-1058967133), if you want a link in the meantime.